client/internal/vanus/store/block_store.go

```diff
@@ -38,32 +42,260 @@ import (
+		return nil, err
+	}
+	_, err = s.connectReadStream(context.Background())
+	if err != nil {
+		// close append stream
+		s.append.releaseStream()
+		// close client
+		s.client.Close()
 		// TODO: check error
 		return nil, err
 	}
 	return s, nil
 }
 
+type streamState string
+
+var (
+	stateRunning streamState = "running"
+	stateClosed  streamState = "closed"
+)
+
+type appendStreamCache struct {
+	opaqueID  uint64
+	stream    segpb.SegmentServer_AppendToBlockStreamClient
+	callbacks sync.Map
+	state     streamState
+	once      sync.Once
+}
+
+func (a *appendStreamCache) isRunning() bool {
+	return a.state == stateRunning
+}
+
+func (a *appendStreamCache) isClosed() bool {
+	return a.state == stateClosed
+}
+
+func (a *appendStreamCache) release() {
+	a.releaseStream()
+	a.releaseCallbacks()
+}
+
+func (a *appendStreamCache) releaseStream() {
+	a.once.Do(func() {
+		a.stream.CloseSend()
+		a.state = stateClosed
+	})
+}
+
+func (a *appendStreamCache) releaseCallbacks() {
+	a.callbacks.Range(func(key, value interface{}) bool {
+		value.(appendCallback)(&segpb.AppendToBlockStreamResponse{
+			ResponseCode: errpb.ErrorCode_CLOSED,
+			ResponseMsg:  "append stream closed",
+			Offsets:      []int64{},
+		})
+		return true
+	})
+}
+
+type readStreamCache struct {
+	opaqueID  uint64
+	stream    segpb.SegmentServer_ReadFromBlockStreamClient
+	callbacks sync.Map
+	state     streamState
+	once      sync.Once
+}
+
+func (r *readStreamCache) isRunning() bool {
+	return r.state == stateRunning
+}
+
+func (r *readStreamCache) isClosed() bool {
+	return r.state == stateClosed
+}
+
+func (r *readStreamCache) release() {
+	r.releaseStream()
+	r.releaseCallbacks()
+}
+
+func (r *readStreamCache) releaseStream() {
+	r.once.Do(func() {
+		r.stream.CloseSend()
+		r.state = stateClosed
+	})
+}
+
+func (r *readStreamCache) releaseCallbacks() {
+	r.callbacks.Range(func(key, value interface{}) bool {
+		value.(readCallback)(&segpb.ReadFromBlockStreamResponse{
+			ResponseCode: errpb.ErrorCode_CLOSED,
+			ResponseMsg:  "read stream closed",
+			Events: &cepb.CloudEventBatch{
+				Events: []*cepb.CloudEvent{},
+			},
+		})
+		return true
+	})
+}
+
 type BlockStore struct {
 	primitive.RefCount
-	client rpc.Client
-	tracer *tracing.Tracer
+	client   rpc.Client
+	tracer   *tracing.Tracer
+	append   *appendStreamCache
+	read     *readStreamCache
+	appendMu sync.RWMutex
+	readMu   sync.RWMutex
+}
+
+type appendCallback func(*segpb.AppendToBlockStreamResponse)
+type readCallback func(*segpb.ReadFromBlockStreamResponse)
+
+func (s *BlockStore) runAppendStreamRecv(ctx context.Context, append *appendStreamCache) {
+	for {
+		res, err := append.stream.Recv()
+		if err != nil {
+			log.Error(ctx, "append stream recv failed", map[string]interface{}{
+				log.KeyError: err,
+			})
+			// only release the remaining callbacks after the stream connection is closed
+			if err == io.EOF || append.isClosed() {
+				append.releaseCallbacks()
+				return
+			}
+			break
```

Comment:
set status to closed.

Reply:
I don't think it is necessary to set the closed status here. The break has exited stream recv